File: app/src/main/java/at/shockbytes/corey/ui/fragment/dialogs/TimeExerciseCountdownDialogFragment.java
package at.shockbytes.corey.ui.fragment.dialogs;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.DialogFragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.concurrent.TimeUnit;

import at.shockbytes.corey.R;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;

/**
 * @author Martin Macheiner
 *         Date: 21.03.2015.
 */
public class TimeExerciseCountdownDialogFragment extends DialogFragment {

    private TextView txtTimer;

    public interface OnCountDownCompletedListener {
        void onCountdownCompleted();
    }

    private static final String ARG_SECONDS = "arg_seconds";

    public static TimeExerciseCountdownDialogFragment newInstance(int seconds) {

        TimeExerciseCountdownDialogFragment fragment = new TimeExerciseCountdownDialogFragment();
        Bundle args = new Bundle(1);
        args.putInt(ARG_SECONDS, seconds);
        fragment.setArguments(args);
        return fragment;
    }

    private int countdown;
    private OnCountDownCompletedListener listener;

    private Disposable disposable = null;

    public TimeExerciseCountdownDialogFragment() {
    }

    public void setCountdownCompleteListener(OnCountDownCompletedListener listener) {
        this.listener = listener;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setCancelable(false);
        setStyle(STYLE_NO_TITLE, 0);
        countdown = getArguments().getInt(ARG_SECONDS);
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        getDialog().getWindow().getAttributes().windowAnimations = R.style.DialogAnimation;
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {

        View v = inflater.inflate(R.layout.dialogfragment_countdown, container, false);
        txtTimer = v.findViewById(R.id.dialogfr_countdown_txt_timer);
        txtTimer.setText(String.valueOf(countdown));
        return v;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initialize();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (disposable != null && !disposable.isDisposed()) {
            disposable.dispose();
        }
    }

    private void initialize() {

        disposable = Observable.interval(1, TimeUnit.SECONDS)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.computation())
                .subscribe(new Consumer<Long>() {
                    @Override
                    public void accept(Long aLong) {

                        countdown--;
                        if (countdown == 0) {
                            if (listener != null) {
                                listener.onCountdownCompleted();
                            }
                            disposable.dispose();
                            dismiss();
                        }
                        txtTimer.setText(String.valueOf(countdown));
                    }
                });
    }
}

File: app/src/main/java/at/shockbytes/corey/ui/fragment/dialog/TimeExerciseCountdownDialogFragment.kt
package at.shockbytes.corey.ui.fragment.dialog

import android.os.Bundle
import androidx.fragment.app.DialogFragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import at.shockbytes.corey.R
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import java.util.concurrent.TimeUnit

/**
 * Author:  Martin Macheiner
 * Date:    21.03.2015
 */
class TimeExerciseCountdownDialogFragment : DialogFragment() {

    private lateinit var txtTimer: TextView

    private var countdown: Int = 0
    private var listener: (() -> Unit)? = null

    private var disposable: Disposable? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        isCancelable = false
        setStyle(DialogFragment.STYLE_NO_TITLE, 0)
        countdown = arguments?.getInt(ARG_SECONDS) ?: 5
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        dialog?.window?.attributes?.windowAnimations = R.style.DialogAnimation
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val v = inflater.inflate(R.layout.dialogfragment_countdown, container, false)
        txtTimer = v.findViewById(R.id.dialogfragment_countdown_txt_timer)
        return v
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupTimer()
    }

    override fun onDestroy() {
        super.onDestroy()
        disposable?.dispose()
    }

    fun setCountdownCompleteListener(listener: (() -> Unit)): TimeExerciseCountdownDialogFragment {
        this.listener = listener
        return this
    }

    private fun setupTimer() {

        txtTimer.text = countdown.toString()

        disposable = Observable.interval(1, TimeUnit.SECONDS)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.computation())
                .subscribe {
                    countdown--
                    if (countdown == 0) {
                        listener?.invoke()
                        disposable?.dispose()
                        dismiss()
                    }
                    txtTimer.text = countdown.toString()
                }
    }

    companion object {

        private const val ARG_SECONDS = "arg_seconds"

        fun newInstance(seconds: Int): TimeExerciseCountdownDialogFragment {
            val fragment = TimeExerciseCountdownDialogFragment()
            val args = Bundle(1)
            args.putInt(ARG_SECONDS, seconds)
            fragment.arguments = args
            return fragment
        }
    }
}
